File: comparison.cpp
#include "comparison.hpp"

int compare(nlohmann::json json_1, nlohmann::json json_2){
    nlohmann::json json_1_only;
    nlohmann::json json_2_only;
    nlohmann::json json_1_more_version;
    ll i = 0, j = 0;
    while (i != json_1["length"] && j != json_2["length"]){
        if (json_2["packages"][j]["arch"] < json_1["packages"][i]["arch"]){
            json_2_only += json_2["packages"][j];
            j++;
            continue;
        }
        if (json_2["packages"][j]["arch"] > json_1["packages"][i]["arch"]){
            json_1_only += json_1["packages"][i];
            i++;
            continue;
        }
        if (json_2["packages"][j]["name"] == json_1["packages"][i]["name"]){
            if (json_2["packages"][j]["version"] < json_1["packages"][i]["version"])
                json_1_more_version += json_1["packages"][i];
            i++;
            j++;
        }else if (json_2["packages"][j]["name"] > json_1["packages"][i]["name"]){
            json_1_only += json_1["packages"][i];
            i++;
        }else{
            json_2_only += json_2["packages"][j];
            j++;
        }
    }
    ofstream out;
    out.open("json_1_only.json");
    if (out.is_open()){
        out<<json_1_only<<endl;
    }else{
        cout<<"Error with saving to a file!"<<endl;
        return 1;
    }
    out.close(); 
    out.open("json_2_only.json");
    if (out.is_open()){
        out<<json_2_only<<endl;
    }else{
        cout<<"Error with saving to a file!"<<endl;
        return 1;
    }
    out.close(); 
    out.open("json_1_more_version.json");
    if (out.is_open()){
        out<<json_1_more_version<<endl;
    }else{
        cout<<"Error with saving to a file!"<<endl;
        return 1;
    }
    out.close(); 
    return 0;
}

int get_response(http::response<http::string_body> *res, string branch){
    string full_path = path + branch;
    io_service svc;
    ssl::context ctx(ssl::context::sslv23_client);
    ssl::stream<ip::tcp::socket> ssocket = {svc, ctx};
    ip::tcp::resolver resolver(svc);
    auto it = resolver.resolve(host, port);
    connect(ssocket.lowest_layer(), it);
    ssocket.handshake(ssl::stream_base::handshake_type::client);
    http::request<http::string_body> req = {http::verb::get, full_path, 11};
    req.set(http::field::host, host);
    write(ssocket, req);
    flat_buffer buffer;
    read(ssocket, buffer, *res);
    boost::system::error_code ec;
    ssocket.shutdown();
    ssocket.lowest_layer().shutdown(ip::tcp::socket::shutdown_both, ec);
    ssocket.lowest_layer().close();
    if (obsolete_reason(res->result()) != "OK"){
        cout<<"Incorrect request! Try again!"<<endl;
        return 1;
    }
    return 0;
}

File: Python/comparison.py
import requests
import json

def compare(branch_1, branch_2):    
    
    response_1 = requests.get('https://rdb.altlinux.org/api/export/branch_binary_packages/' + branch_1)
    if (response_1.status_code != 200):
        print("Incorrect request! Try again!")
        return
    
    response_2 = requests.get('https://rdb.altlinux.org/api/export/branch_binary_packages/' + branch_2)
    if (response_2.status_code != 200):
        print("Incorrect request! Try again!")
        return
    
    json_1 = response_1.json()
    json_2 = response_2.json()
    
    i = 0
    j = 0
    json_1_only = []
    json_2_only = []
    json_1_more_version = []
    
    while (i != json_1["length"] and j != json_2["length"]):
        if (json_2["packages"][j]["arch"] < json_1["packages"][i]["arch"]):
            json_2_only.append(json_2["packages"][j])
            j = j + 1
            continue
        if (json_2["packages"][j]["arch"] > json_1["packages"][i]["arch"]):
            json_1_only.append(json_1["packages"][i])
            i = i + 1
            continue
        if (json_2["packages"][j]["name"] == json_1["packages"][i]["name"]):
            if (json_2["packages"][j]["version"] < json_1["packages"][i]["version"]):
                json_1_more_version.append(json_1["packages"][i])
            i = i + 1
            j = j + 1
        elif (json_2["packages"][j]["name"] > json_1["packages"][i]["name"]):
            json_1_only.append(json_1["packages"][i])
            i = i + 1
        else:
            json_2_only.append(json_2["packages"][j])
            j = j + 1
    
    with open('json_1_only.json', 'w') as file:
        json.dump(json_1_only, file)
    file.close()

    with open('json_2_only.json', 'w') as file:
        json.dump(json_2_only, file)
    file.close()

    with open('json_1_more_version.json', 'w') as file:
        json.dump(json_1_more_version, file)
    file.close()